File: exercises/python_playwright/Visibility/visibility_playwright.py
from playwright.sync_api import Error, Locator, Page


class VisibilityPage:
    """Page object for the Visibility exercise."""

    def __init__(self, page: Page):
        """Initialize the page and button locators."""
        self.page = page
        self.hide_button = page.get_by_role("button", name="Hide")
        self.removed_button = page.locator("#removedButton")
        self.zero_width_button = page.locator("#zeroWidthButton")
        self.overlapped_button = page.locator("#overlappedButton")
        self.zero_opacity_button = page.locator("#transparentButton")
        self.hidden_button = page.locator("#invisibleButton")
        self.none_button = page.locator("#notdisplayedButton")
        self.offscreen_button = page.locator("#offscreenButton")

    def open(self) -> None:
        """Open the Visibility exercise page."""
        self.page.goto("http://uitestingplayground.com/visibility")

    def click_hide_button(self) -> None:
        """Click the button that hides the target buttons."""
        self.hide_button.click()

    def is_invisible_for_user(self, locator: Locator) -> tuple[bool | str]:
        """Return invisibility status and the first matching reason."""
        # built in is_visible() check
        if not locator.is_visible():
            return True, "Playwright invisible"

        # opacity check
        opacity = float(locator.evaluate("el => getComputedStyle(el).opacity"))
        if opacity == 0:
            return True, "Opacity = 0"

        # offscreen check
        offscreen = locator.evaluate(
        """
        el => {
            const r = el.getBoundingClientRect();
            return (
                r.bottom < 0 ||
                r.right < 0 ||
                r.top > window.innerHeight ||
                r.left > window.innerWidth
            );
        }
        """)
        if offscreen:
            return True, "Button Offscreen"

        # not clickable by user
        try:
            locator.click(trial=True, timeout=2000)
        except Error:
            return True, "Button Overlapped"

        # passed all checks
        return False, "Button visible to users"

File: exercises/javascript_playwright/visibility/visibilityPlaywright.js
export class VisibilityPage {
    /**
     *
     * @param {import("@playwright/test").Page} page
     */
    constructor(page) {
        this.page = page;
        this.hideButton = page.locator("#hideButton");
        this.removedButton = page.locator("#removedButton");
        this.zeroWidthButton = page.locator("#zeroWidthButton");
        this.overlappedButton = page.locator("#overlappedButton");
        this.zeroOpacityButton = page.locator("#transparentButton");
        this.hiddenButton = page.locator("#invisibleButton");
        this.noneButton = page.locator("#notdisplayedButton");
        this.offscreenButton = page.locator("#offscreenButton");
    }

    async openExercisePage() {
        await this.page.goto("http://uitestingplayground.com/visibility");
    }

    async clickHideButton() {
        await this.hideButton.click();
    }

    async checkIfButtonIsInvisibleToUser(button) {
        // built in method
        if (!(await button.isVisible())) {
            return true;
        }

        // check opacity
        const opacity = await button.evaluate((el) => {
            return window.getComputedStyle(el).opacity;
        });
        if (opacity === "0") {
            return true;
        }

        // check if overlapped
        const overlapped = await button.evaluate((el) => {
            const rect = el.getBoundingClientRect();
            const centerX = rect.left + rect.width / 2;
            const centerY = rect.top + rect.height / 2;
            const elementAtPoint = document.elementFromPoint(centerX, centerY);
            return !el.contains(elementAtPoint) && elementAtPoint !== el;
        });
        if (overlapped) {
            return true;
        }

        // check if offscreen
        const offscreen = await button.evaluate((el) => {
            const rect = el.getBoundingClientRect();
            return (
                rect.bottom < 0 ||
                rect.right < 0 ||
                rect.left > window.innerWidth ||
                rect.top > window.innerHeight
            );
        });
        if (offscreen) {
            return true;
        }

        return false;
    }
}
